tests/msc2716_test.go

```diff
@@ -1211,3 +1254,159 @@ func batchSendHistoricalMessages(
+					return r.Get("event_id").Str
+				}, nil),
+			},
+		})
+	}
+
+	// Make sure the historical events appear in scrollback without jumping back
+	// in time specifically.
+	fullMessagesRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+		"dir":   []string{"b"},
+		"limit": []string{"100"},
+	}))
+	must.MatchResponse(t, fullMessagesRes, match.HTTPResponse{
+		JSON: []match.JSON{
+			matcherJSONEventIDArrayInOrder("chunk",
+				expectedEventIDOrder,
+				historicalEventFilter,
+			),
+		},
+	})
+}
+
+// Looks through a list of events to find the sliding window of expected event
+// ID's somewhere in the list in order. The expected list can start anywhere in
+// the overall list.
+func matcherJSONEventIDArrayInOrder(wantKey string, expectedEventIDOrder []string, eventFilter func(gjson.Result) bool) match.JSON {
+	return func(body []byte) error {
+		if len(expectedEventIDOrder) == 0 {
+			return fmt.Errorf("expectedEventIDOrder can not be an empty list")
+		}
+
+		// Copy the array by slice so we can modify it as we iterate in the foreach loop.
+		// We save the full untouched `expectedEventIDOrder` for use in the log messages
+		workingExpectedEventIDOrder := expectedEventIDOrder
+
+		var res gjson.Result
+		if wantKey == "" {
+			res = gjson.ParseBytes(body)
+		} else {
+			res = gjson.GetBytes(body, wantKey)
+		}
+
+		if !res.Exists() {
+			return fmt.Errorf("missing key '%s'", wantKey)
+		}
+		if !res.IsArray() {
+			return fmt.Errorf("key '%s' is not an array", wantKey)
+		}
+
+		eventDebugStringsFromResponse, err := _getRelevantEventDebugStringsFromMessagesResponse("chunk", body, eventFilter)
+		if err != nil {
+			return err
+		}
+
+		// Loop through the overall event list
+		foundFirstEvent := false
+		res.ForEach(func(_, r gjson.Result) bool {
+			eventID := r.Get("event_id").Str
+			nextEventIdInOrder := workingExpectedEventIDOrder[0]
+
+			// We need to find the start of the sliding window inside the overall
+			// event list
+			if !foundFirstEvent && eventID == nextEventIdInOrder {
+				foundFirstEvent = true
+			}
+
+			// Once we found the first event, find all events in order
+			if foundFirstEvent && eventFilter(r) {
+				if r.Get("event_id").Str != nextEventIdInOrder {
+					err = fmt.Errorf("Next event found was %s but expected %s\nActualEvents (%d): %v\nExpectedEvents (%d): %v", r.Get("event_id").Str, nextEventIdInOrder, len(eventDebugStringsFromResponse), eventDebugStringsFromResponse, len(expectedEventIDOrder), expectedEventIDOrder)
+				}
+
+				// Now that we found it, pop the message off the expected list
+				workingExpectedEventIDOrder = workingExpectedEventIDOrder[1:]
+			}
+
+			// Found all of the expected events, stop iterating
+			if len(workingExpectedEventIDOrder) == 0 {
+				return false
+			}
+
+			return err == nil
```

Comment:
Control flow feels weird here, why not just assign to `err` and `return false` up at `:1389`?

Reply:
That works, updated 👍

For reference, it was that way because it was mimicking [`JSONArrayEach`](https://github.com/matrix-org/complement/blob/5535399363421bc45b762dce3d4d8432607ce583/internal/match/json.go#L182) and I previously had this middle logic refactored into an internal callback at some point.